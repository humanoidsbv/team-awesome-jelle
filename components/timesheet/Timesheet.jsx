import React from 'react';
import PropTypes from 'prop-types';

import TimesheetDate from '../timesheet-date/TimesheetDate';
import TimesheetEntry from '../timesheet-entry/TimesheetEntry';
import TimesheetInput from '../timesheet-input/TimesheetInput';

import { fetchTimesheetEntries, postTimesheetEntry, deleteTimesheetEntry } from '../../services/fetch-timesheet-entries/fetch-timesheet-entries';

import './timesheet.scss';

class Timesheet extends React.Component {
  static propTypes = {
    timesheetEntries: PropTypes.arrayOf(PropTypes.shape({
      employer: PropTypes.string.isRequired,
      id: PropTypes.string.isRequired,
      startTime: PropTypes.string.isRequired,
      endTime: PropTypes.string.isRequired,
      date: PropTypes.string.isRequired
    })).isRequired,
    onDeleteTimesheetEntry: PropTypes.func.isRequired,
    onDeleteTimesheetEntrySucces: PropTypes.func.isRequired,
    onPostTimesheetEntry: PropTypes.func.isRequired,
    onPostTimesheetEntrySucces: PropTypes.func.isRequired,
    onRequestTimeEntries: PropTypes.func.isRequired,
    onRequestTimeEntriesSucces: PropTypes.func.isRequired
  }

  componentDidMount() {
    const { onRequestTimeEntries, onRequestTimeEntriesSucces } = this.props;
    onRequestTimeEntries();
    fetchTimesheetEntries().then((timesheetEntries) => {
      onRequestTimeEntriesSucces(timesheetEntries);
    });
  }

  handleAddTimesheetEntry = (newEntry) => {
    const { onPostTimesheetEntry, onPostTimesheetEntrySucces } = this.props;
    onPostTimesheetEntry();
    postTimesheetEntry(newEntry).then((postedEntry) => {
      onPostTimesheetEntrySucces(postedEntry);
    });
  };

  handleDeleteTimesheetEntry = (timesheetEntryId) => {
    const { onDeleteTimesheetEntry, onDeleteTimesheetEntrySucces } = this.props;
    onDeleteTimesheetEntry();
    deleteTimesheetEntry(timesheetEntryId).then(() => {
      onDeleteTimesheetEntrySucces(timesheetEntryId);
    });
  }


  render() {
    const { timesheetEntries } = this.props;
    return (
      <div className="timesheet-wrapper">
        <TimesheetInput
          onSave={this.handleAddTimesheetEntry}
        />
        {timesheetEntries.map((timesheetEntry, index, array) => (
          <React.Fragment key={timesheetEntry.id}>
            {(!index || (timesheetEntry.date !== array[index - 1].date)) && (
            <TimesheetDate
              date={timesheetEntry.date}
            />
            )}
            <TimesheetEntry
              employer={timesheetEntry.employer}
              endTime={timesheetEntry.endTime}
              id={timesheetEntry.id}
              onDelete={this.handleDeleteTimesheetEntry}
              startTime={timesheetEntry.startTime}
            />
          </React.Fragment>
        ))}
      </div>
    );
  }
}

export default Timesheet;
