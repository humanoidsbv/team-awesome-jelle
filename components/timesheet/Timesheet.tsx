import React from 'react';

import TimesheetSearchBar from '../timesheet-search-bar/TimesheetSearchBar';
import TimesheetDate from '../timesheet-date/TimesheetDate';
import TimesheetEntry from '../timesheet-entry/TimesheetEntry';
import AddTimesheet from '../add-timesheet/AddTimesheet';

import { calculateCumulativeDuration } from '../../services/convert-time/convert-time';

import { TimesheetEntryModel } from '../../ducks/time-entries';
import { ClientOptionModel } from '../../ducks/clients'

import './timesheet.scss';

interface TimesheetProps {
  activeFilter: string;
  timesheetEntries: TimesheetEntryModel[];
  clientOptions: ClientOptionModel[];
  isFormSaving: boolean;
  onDeleteTimesheetEntry: (id: number) => void;
  onPostTimesheetEntry: (newTimesheet: TimesheetEntryModel[]) => void;
  onRequestTimeEntries: () => void;
  onChangeActiveFilter: (newActiveFilter: string) => void;
}

class Timesheet extends React.Component<TimesheetProps> {
  componentDidMount() {
    this.props.onRequestTimeEntries();
  }

  handleAddTimesheetEntry = newEntry => this.props.onPostTimesheetEntry(newEntry);

  handleDeleteTimesheetEntry = (timesheetEntryId) => {
    this.props.onDeleteTimesheetEntry(timesheetEntryId);
  }

  render(){
    const {
      timesheetEntries, isFormSaving,
      onChangeActiveFilter, activeFilter,
      clientOptions
    } = this.props;

    return (
      <div className="timesheet">
        <TimesheetSearchBar
          onChangeActiveFilter={onChangeActiveFilter}
          activeFilter={activeFilter}
          clientOptions={clientOptions}
        />
        <div className="timesheet__wrapper">
          <AddTimesheet
            clientOptions={clientOptions}
            onSave={this.handleAddTimesheetEntry}
            isFormSaving={isFormSaving}
          />
          <ul>
            {timesheetEntries.map((timesheetEntry, index, array) => (
              <React.Fragment key={timesheetEntry.id}>
                {(!index || (timesheetEntry.date !== array[index - 1].date)) && (
                <li>
                  <TimesheetDate
                    date={timesheetEntry.date}
                    totalTime={calculateCumulativeDuration(
                      timesheetEntry.startTime, timesheetEntries
                    )}
                  />
                </li>
                )}
                <li>
                  <TimesheetEntry
                    timesheet={timesheetEntry}
                    onDelete={this.handleDeleteTimesheetEntry}
                  />
                </li>
              </React.Fragment>
            ))}
          </ul>
        </div>
      </div>
    );
  }
}

export default Timesheet;
