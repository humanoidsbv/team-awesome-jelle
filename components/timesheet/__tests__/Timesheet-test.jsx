import React from 'react';
import { shallow } from 'enzyme';

import Timesheet from '../Timesheet';

const deleteTimesheetEntry = () => {};
const requestTimeEntries = () => {};
const postTimesheetEntry = () => {};

const timesheetEntriesSelector = [
  {
    employer: 'Port of Rotterdam',
    activity: 'Design',
    date: '2018-12-13',
    startTime: '2018-12-13T18:00:00.000Z',
    endTime: '2018-12-13T18:30:00.000Z',
    id: 1
  },
  {
    employer: 'Port of Rotterdam',
    activity: 'Design',
    date: '2012-12-13',
    startTime: '2012-12-13T12:00:00.000Z',
    endTime: '2012-12-13T13:00:00.000Z',
    id: 2
  }
];

const isFormSavingSelector = false;

it('Renders Date Component correctly', () => {
  const wrapper = shallow(
    <Timesheet
      isFormSaving={isFormSavingSelector}
      onDeleteTimesheetEntry={deleteTimesheetEntry}
      onPostTimesheetEntry={postTimesheetEntry}
      onRequestTimeEntries={requestTimeEntries}
      timesheetEntries={timesheetEntriesSelector}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
