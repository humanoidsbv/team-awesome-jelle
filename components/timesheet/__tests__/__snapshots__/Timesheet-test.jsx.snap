// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Date Component correctly 1`] = `
<div
  className="timesheet__page-wrapper"
>
  <TimesheetSearchBar />
  <div
    className="timesheet__wrapper"
  >
    <TimesheetInput
      isFormSaving={false}
      onSave={[Function]}
    />
    <TimesheetDate
      date="2018-12-13"
    />
    <TimesheetEntry
      employer="Port of Rotterdam"
      endTime="2018-12-13T18:30:00.000Z"
      id={1}
      onDelete={[Function]}
      startTime="2018-12-13T18:00:00.000Z"
    />
    <TimesheetDate
      date="2012-12-13"
    />
    <TimesheetEntry
      employer="Port of Rotterdam"
      endTime="2012-12-13T13:00:00.000Z"
      id={2}
      onDelete={[Function]}
      startTime="2012-12-13T12:00:00.000Z"
    />
  </div>
</div>
`;
